scheduler/scheduler.go

```diff
@@ -0,0 +1,117 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scheduler
+
+import (
+	"context"
+	"time"
+
+	"github.com/getgort/gort/adapter"
+	"github.com/getgort/gort/command"
+	"github.com/getgort/gort/data"
+	"github.com/getgort/gort/dataaccess"
+	"github.com/getgort/gort/telemetry"
+	"github.com/go-co-op/gocron"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/codes"
+)
+
+type CommandScheduler struct {
+	cron     *gocron.Scheduler
+	Commands chan data.CommandRequest
+	da       dataaccess.DataAccess
+}
+
+func NewCommandScheduler(da dataaccess.DataAccess) CommandScheduler {
+	cron := gocron.NewScheduler(time.Local)
+	cron.TagsUnique()
+	return CommandScheduler{
+		cron:     cron,
+		Commands: make(chan data.CommandRequest),
+		da:       da,
+	}
+}
+
+func (cs *CommandScheduler) Start() {
+	cs.cron.StartAsync()
+}
+
+func (cs *CommandScheduler) schedule(ctx context.Context, command data.ScheduledCommand) error {
+	err := cs.da.ScheduleCreate(ctx, &command)
+	if err != nil {
+		return err
+	}
+	_, err = cs.cron.Cron(command.Cron).Do(func() { //todo tag with scheduleid
+		tr := otel.GetTracerProvider().Tracer(telemetry.ServiceName)
+		ctx, sp := tr.Start(context.Background(), "adapter.handleIncomingEvent")
+		defer sp.End()
+
+		req := data.CommandRequest{
+			CommandEntry: command.CommandEntry,
+			Adapter:      command.Adapter,
+			ChannelID:    command.ChannelID,
+			Context:      ctx,
+			Parameters:   command.Parameters,
+			Timestamp:    time.Now(),
+			UserID:       command.UserID,
+			UserEmail:    command.UserEmail,
+			UserName:     command.UserName,
+		}
+
+		err := cs.da.RequestBegin(ctx, &req)
+		if err != nil {
+			sp.RecordError(err)
+			sp.SetStatus(codes.Error, "Failed to begin request")
+			return
+		}
+
+		cs.Commands <- req
+	})
+
+	return err
+}
+
+func (cs *CommandScheduler) Add(ctx context.Context, cron string, cmdEntry data.CommandEntry, parameters data.CommandParameters, userID, userEmail, userName, adapterName, channelID string) error {
+
+	err := cs.schedule(ctx, data.ScheduledCommand{
+		CommandEntry: cmdEntry,
+		Adapter:      adapterName,
+		ChannelID:    channelID,
+		Parameters:   parameters,
+		UserID:       userID,
+		UserEmail:    userEmail,
+		UserName:     userName,
+		Cron:         cron,
+	})
+
+	return err
+}
+
+func (cs *CommandScheduler) AddFromString(ctx context.Context, cron, commandString, userID, userEmail, userName, adapterName, channelID string) error {
```

Comment:
This is a lot of strings in a row, which makes the function hard to use correctly without closely reading the documentation. I think you might want to consider something like function currying to replace both `Add` functions. Something like:

```go
type AddOption func(c *data.ScheduledCommand) error

func (cs *CommandScheduler) Add(ctx context.Context, options ...AddOption) error {
   var s data.ScheduledCommand

   for _, o := range options {
      if err := o(&s); err != nil {
         return fmt.Errorf("error setting option: %w")
      }

   return cs.schedule(ctx, s)
}

// This allows you to have multiple With* functions that accepts 
// different flavors of command

func WithCommandEntry(ce command.CommandEntry) AddOption {
   return func(c *data.ScheduledCommand) error {
      c.CommandEntry = ce
      return nil
   }
}

func WithCommandString(commandString string) AddOption {
   return func(c *data.ScheduledCommand) error {
      tokens, err := command.Tokenize(commandString)
      if err != nil {
         return err
      }

      cmdEntry, cmdInput, err := adapter.CommandFromTokensByName(ctx, tokens)
      if err != nil {
         return err
      }

      c.CommandEntry = *cmdEntry
      c.Parameters = adapter.ParametersFromCommand(cmdInput)

      return nil
   }
}
```

I wrote this freehand, but it should work okay and hopefully is enough to get you started.